contents/handbook/engineering/async-migrations.md

```diff
@@ -0,0 +1,222 @@
+---
+title: Async Migrations
+sidebar: Handbook
+showTitle: true
+---
+
+> **Note:** Async migrations are currently a work in progress under the name of "special migrations".
+
+## What are async migrations?
+
+Async migrations (AKA big/special migrations) are migrations that do not run synchronously on an update to a PostHog instance. Rather, they execute on the background of a running PostHog instance, and should be completed within a range of PostHog versions. 
+
+## Why are async migrations necessary?
+
+Migrations are inevitable, and sometimes it may be necessary to execute non-trivial schema changes that can take a long time to complete. 
+
+For example, ClickHouse does not support changing the primary key of a table, which is a change we were [forced to make in anticipation of upgrading ClickHouse beyond version 21.6](https://github.com/PostHog/posthog/issues/5684). As a result, the way to change the schema of the table was to create a new table and insert all the data from the old table into it, which took us an entire week to run on PostHog Cloud.
+
+Now, while we can execute such changes to PostHog Cloud "manually" (i.e. James G SSHs into CH5, runs a bunch of queries using clickhouse-client, and keeps an eye on them), we cannot expect the same of our users (plus this approach does not scale well for us either).
+
+As a result, we needed to create a system capable of safely and efficiently managing migrations that need to happen asynchronously.
+
+## Why did we write this system from scratch?
+
+As a provider of a product that users can self-host, we needed a bespoke solution that interoperates well with our particular tech stack and requirements, such as:
+
+1. The ability to run asynchronous tasks using a system already in use (Celery)
```

Comment:
It kind of is. We're reluctant (rightfully so) to introduce new services at PostHog. It's in the handbook somewhere but can't find it rn. Adding another system for this when we already have a worker would be best to avoid although not impossible